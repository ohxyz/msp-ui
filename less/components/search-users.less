@import "./search-box";
@import "./user-list";
