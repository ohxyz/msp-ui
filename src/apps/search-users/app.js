const React = require( 'react' );
const ReactDOM = require( 'react-dom' );
const SearchUsers = require( '../../components/search-users/search-users.js' ).SearchUsers;
const HierarchyStorage = require( '../../components/core/hierarchy-storage.js' ).HierarchyStorage;
const appManager = require( '../core/app-manager.js' ).appManager;

class SearchUsersApp { 

    constructor( { id, sapData, domElement } ) {

        this.sapData = sapData;
        this.domElement = domElement;
        this.hierarchyStorage = new HierarchyStorage( sapData );
        this.searchUsersComponent = <SearchUsers storage={ this.hierarchyStorage } />;

        if ( typeof id === 'string' ) {

            appManager.addApp( id, this );
        }

    }

    reload( sapData ) {

        this.hierarchyStorage = new HierarchyStorage( sapData );
        this.searchUsersComponent = <SearchUsers storage={ this.hierarchyStorage } />;
        this.unload();
        this.load();
    }

    load() {

        ReactDOM.render( this.searchUsersComponent, this.domElement );
    }

    unload() {

        return ReactDOM.unmountComponentAtNode( this.domElement );
    }

    /* 
     * @param user object {
     *
     *      hierarchyId: '"00215A9C-537A-1ED8-91D7-9D2EA1B25283"
     *      firstName: "Tom",
     *      lastName: "Lu"
     *      emailAddress: "tom.lu@tom.lu"
     * }
     *
     */
    addUser( user ) {

        this.hierarchyStorage.addUser( user );
    }
}

module.exports = {

    SearchUsersApp
};