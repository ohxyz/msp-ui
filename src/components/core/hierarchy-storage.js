/* Store and process users data from SAP
 *
 */

const util = require( '../../helpers/util.js' );
const AccountProfile = require( '../core/account-profile.js' ).AccountProfile;
const HierarchyNode = require( '../core/hierarchy-node.js' ).HierarchyNode;

class HierarchyStorage {

    constructor( sapRaw, shouldProcess = false ) {

        this.sapRaw = sapRaw;
        this.sapResults = null;
        this.accounts = []; // users + orgs
        this.nodes = []; // HierarchyNodes
        this.isProcessed = false;

        /*
         *
         * @example [
         *
         *     { "CA2B1FE6-D50C-1ED6-B5FC-07B7BC728307": HierarchyNode }
         *     { "CA2B1FE6-D50C-1ED7-BD9E-55E1ACCF4802": HierarchyNode },
         *     { "CA2B1FE6-D50C-1ED8-8AFC-C1EFC84A5801": HierarchyNode },
         *     { "CA2B1FE6-D50C-1ED8-8AFC-D71F61F20303": HierarchyNode }
         * ]
         */
         this.mapOfHierarchyIdAndNode = { };

         if ( shouldProcess === true ) {

            this.process( this.sapRaw );
         }
    }

    /* 
     *
     * @param content can be a JSON string or an object
     */
    validateSapRaw( content ) {

        let sapObject = null;

        if ( util.isObject( content ) ===  true ) {

            sapObject = content;
        }
        else {

            let parsed = null;

            try {

                parsed = JSON.parse( content );
            }
            catch( error ) {

                throw new Error( '[MSP] Failed to parse SAP data. ' + error.message );
            }

            if ( util.isObject( parsed ) === false ) {

                throw new Error( '[MSP] SAP data should be parsed to an object.');
            }

            sapObject = parsed;
        }

        if ( sapObject.hasOwnProperty( 'results' )
                && Array.isArray( sapObject.results ) === true ) {

            this.sapResults = sapObject.results;

        }
        else {

            try {

                this.sapResults = sapObject.d.results;
            }
            catch( error ) {

                throw new Error( '[MSP] SAP data should have "d.results".' );
            }
        }

        if ( Array.isArray( this.sapResults ) === false ) {

            throw new Error( '[MSP] SAP data\'s "d.results" should be an array.' );
        }
    }

    process() {

        this.validateSapRaw( this.sapRaw );

        let lastNode = new HierarchyNode();
        let lastParentNode = new HierarchyNode();
        let hierarchyNodes = this.sapResults.map( result => new HierarchyNode( result ) );

        for ( let i = 0; i < hierarchyNodes.length; i ++ ) {

            let currentNode = hierarchyNodes[ i ];
            let hierarchyId = currentNode.hierarchyId;

            if ( currentNode.parentId === lastNode.hierarchyId ) {

                currentNode.parent = lastNode;
            }
            else if ( lastNode.parent !== null
                    && lastNode.parent.hierarchyId === currentNode.parentId ) {

                currentNode.parent = lastNode.parent;
            }
            else {

                let parentId = currentNode.parentId;

                if ( this.mapOfHierarchyIdAndNode.hasOwnProperty( parentId ) ) {

                    let parent = this.mapOfHierarchyIdAndNode[ parentId ];
                    currentNode.parent = parent;
                }
            }

            for ( let user of currentNode.users ) {

                if ( currentNode.parent !== null  ) {

                }
            }

            this.accounts.push( ...currentNode.accounts );

            this.mapOfHierarchyIdAndNode[ hierarchyId ] = currentNode;
            this.nodes.push( currentNode );

            lastNode = currentNode;
        }

        this.isProcessed = true;
    }

    getUsersByHierarchyId( hierarchyId, untilLevel = 1 ) {

        let node = this.mapOfHierarchyIdAndNode[ hierarchyId ];

        if ( node === undefined ) {

            return [];
        }

        let users = node.users;
        let topOrg = null;
        let topOrgName = '';
        let levelLiteral = untilLevel.toString();

        while ( node.parent !== null && node.level !== levelLiteral ) {

            let parent = node.parent;

            if ( parent.orgs.length > 0 ) {

                let parentOrg = parent.orgs[ 0 ];

                topOrg = parentOrg;
                topOrgName = parentOrg.name;
            }
            
            users.push( ...parent.users );
            node = parent;
        }

        for ( let user of users ) {

            user.topOrg = topOrg;
            user.topOrgName = topOrgName;
        }

        users = Array.from( new Set( users ) );

        return users;
    }

    getAllUsers() {

        if ( this.isProcessed === false ) {

            this.process();
        }

        let allUsers = [];

        for ( let node of this.nodes ) {

            let hierarchyId = node.hierarchyId;
            let users = this.getUsersByHierarchyId( hierarchyId );

            allUsers.push( ...users );
        }

        allUsers = Array.from( new Set( allUsers ) );

        return allUsers;
    }

    addUser( { hierarchyId, firstName, lastName, emailAddress } ) {

        let node = this.mapOfHierarchyIdAndNode[ hierarchyId ];
        let account = new AccountProfile( { 

            hierarchyId: hierarchyId,
            firstName: firstName,
            lastName: lastName,
            emailAddress: emailAddress,
            orgs: node.orgs
        } );

        if ( node === undefined ) {

            throw new Error( '[MSP] Failed to find Hierarchy ID when adding a user.' );
        }

        node.accounts.push( account );
        node.users.push( account );
        this.accounts.push( account );

        return account;
    }
}


module.exports = {

    HierarchyStorage,
};