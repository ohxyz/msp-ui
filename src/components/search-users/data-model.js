/* Store and process users data from SAP
 *
 */

class UserStorage {

    constructor( sapRaw ) {

        this.sapData = sapRaw;
        this.users = [];
        this.orgs = [];
    }

}

module.exports = {

    UserStorage,
};