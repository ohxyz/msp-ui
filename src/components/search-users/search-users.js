import React from 'react';
import util from '../../helpers/util.js';
import { SearchBox } from '../search-box/search-box.js';
import { UserList } from '../user-list/user-list.js';
import { MessageBox } from '../message-box/message-box.js';
import { componentManager } from '../core/component-manager.js';

const SECONDS_TO_DISMISS_MESSAGE_BOX = 10;
const SUCCESS_CODE = '200';
const ERROR_CODE_4XX = '4xx';
const ERROR_CODE_5XX = '5xx';

class SearchUsers extends React.Component {

    constructor( props ) {

        super( props );

        this.handleSearchBoxItemSelect = this.handleSearchBoxItemSelect.bind( this );
        this.handleTextChange = this.handleTextChange.bind( this );
        this.handleDeleteUserPromise = this.handleDeleteUserPromise.bind( this );

        this.numberOfStrikes = 3;

        this.state = {

            storage: null,
            usersFound: [],
            userToDelete: null,
            shouldRenderUserList: false,
            shouldRenderMessageBox: false,
            messageCode: '',
            onPropsDeleteUser: new Function()
        };

        this.id = util.setDefault( props.id, util.generateRandomString() );
        componentManager.addComponent( this.id, this );
    }

    static getDerivedStateFromProps( nextProps, prevState ) {

        let storage = nextProps.storage;

        if ( storage === undefined ) {

            throw new Error( '[MSP] SearchUsers component must have a HierarchyStorage instance as property.' );
        }

        storage.process();

        return {

            storage: storage,
            onPropsDeleteUser: util.setDefault( nextProps.onDeleteUser, new Function() )
        };
    }

    showAllUsers() {

        let allUsers = this.state.storage.getAllUsers();

        this.setState( {

            usersFound: allUsers,
            shouldRenderUserList: true

        } );
    }

    handleSearchBoxItemSelect( item, searchBox ) {

        let account = item.__origin__;
        let users = [];

        if ( account.type === 'person' ) {

            users = [ account ];
        }
        else {
            
            users = this.state.storage.getUsersByHierarchyId( account.hierarchyId );
        }
        
        this.setState( { 

            usersFound: users,
            shouldRenderUserList: true

        } );
    }

    handleTextChange( usersFound, searchBox ) {

        if ( searchBox.text.length <= this.numberOfStrikes 
                && searchBox.text.length > 0 ) {

            return;
        }
        else if ( searchBox.text.length === 0 ) {

            this.showAllUsers();
            
            return;
        }
    }

    renderSearchBox() {

        return (

            <SearchBox
                id="seach-users-box"
                inputId="search-users-field"
                items={ this.state.storage.accounts }
                fields={ [ 'fullName' ] }
                indexOfFieldsToSort={ 0 }
                strikes={ this.numberOfStrikes }
                placeholder="Search users by name, department or agency"
                onSelect={ this.handleSearchBoxItemSelect }
                onIconClick={ self => { self.clearSearch(); this.showAllUsers(); } }
                onChange={ this.handleTextChange }
            />
        );
    }

    handleDeleteUserPromise( user ) {

        console.log( '*', user.accountId, user.hierarchyId );

        let promise = this.state.onPropsDeleteUser( user );

        promise
            .then( () => {

                this.setState( { 

                    shouldRenderMessageBox: true,
                    messageCode: SUCCESS_CODE,
                    userToDelete: user
                } );
                user.xxxxxxxxxxxxxxxxxx = 1;
                return user;

            } )
            .catch( errorCode => {

                let errorCodeLiteral = errorCode.toString();
                let messageCode = '';

                if ( errorCodeLiteral[ 0 ] === '4' ) {

                    messageCode = ERROR_CODE_4XX;
                }
                else if ( errorCodeLiteral[ 0 ] === '5' ) {

                    messageCode = ERROR_CODE_5XX;
                }

                this.setState( { 

                    shouldRenderMessageBox: true,
                    messageCode: messageCode,
                    userToDelete: user
                } );

            } );

        return promise;
    }

    renderMessageBox( messageCode, user ) {

        let type = '';
        let title = '';
        let content = '';

        if ( messageCode === SUCCESS_CODE ) {

            type = 'success';
            title = 'Deletion achieved';
            content = <React.Fragment>You've deleted the user account for <em>{ user.fullName }</em>.</React.Fragment>;
        }
        else if ( messageCode === ERROR_CODE_4XX ) {

            type = 'error';
            title = 'Oops. Delete user failed.';
            content = (

                <React.Fragment>
                    Sorry - something's gone wrong in deleting this user. Please contact your <em>Origin Account Manager</em>.
                </React.Fragment>
            );
        }
        else if ( messageCode === ERROR_CODE_5XX ) {

            type = 'error';
            title = 'Oops. Delete user failed.';
            content = "Sorry - something's gone wrong in deleting this user account. Please try again later.";
        }
        else {

            return null;
        }

        return (

            <MessageBox 
                type={ type } 
                title={ title } 
                content={ content } 
                seconds={ SECONDS_TO_DISMISS_MESSAGE_BOX } 
            />
        );
    }

    renderUserList() {

        if( this.state.shouldRenderUserList === false ) {

            return;
        }

        return (

            <UserList
                id="search-users-list"
                users={ this.state.usersFound }
                onRenderCount={ count => `${count} users found.` }
                sortByFields={ [ 'firstName', 'lastName' ] }
                onDeleteUser={ this.handleDeleteUserPromise }
            />
        );
    }

    render() {

        let userToDelete = this.state.userToDelete;
        let messageCode = this.state.messageCode;

        let shouldRenderMessageBox = this.state.shouldRenderMessageBox === true 
                                        && messageCode !== '' 
                                        && userToDelete !== null;

        let messageBox = shouldRenderMessageBox === true 
                       ? this.renderMessageBox( messageCode, userToDelete )
                       : null;
        return (

            <div id="search-users-section">
                { messageBox }
                { this.renderSearchBox() }
                { this.renderUserList() }
            </div>
        );
    }
}

module.exports = {

    SearchUsers
};