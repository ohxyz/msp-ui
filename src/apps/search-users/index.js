import React from 'react';
import ReactDOM from 'react-dom';
import { SearchUsers } from '../../components/search-users/search-users.js';
import { HierarchyStorage } from '../../components/core/hierarchy-storage.js';
import { appManager } from '../core/app-manager.js';

class SearchUsersApp { 

    constructor( { id, sapData, domElement } ) {

        this.sapData = sapData;
        this.domElement = domElement;
        this.hierarchyStorage = new HierarchyStorage( sapData );
        this.searchUsersReactElement = <SearchUsers storage={ this.hierarchyStorage } onDeleteUser={ this.deleteUser } />;
        this.searchUsersReactComponent = null;

        if ( typeof id === 'string' ) {

            appManager.addApp( id, this );
        }
    }

    deleteUser( user ) {

        let httpStatusCode = 200;
        let httpStatusCodeLiteral = httpStatusCode.toString();

        // Mock
        let promise = new Promise( ( resolve, reject ) => {

            if ( httpStatusCodeLiteral === '200' ) {

                setTimeout( () => resolve( user ), 2000 );
            }
            else { 

                setTimeout( () => { 

                    reject( new Error( httpStatusCodeLiteral ) );

                }, 2000 );
            }

        } );

        return promise;
    }

    reload( sapData ) {

        this.hierarchyStorage = new HierarchyStorage( sapData );
        this.searchUsersReactElement = <SearchUsers storage={ this.hierarchyStorage } onDeleteUser={ this.deleteUser } />;
        this.unload();
        this.load();
    }

    load() {

        this.searchUsersReactComponent = ReactDOM.render( this.searchUsersReactElement, this.domElement );
        this.searchUsersReactComponent.showAllUsers();
    }

    unload() {

        this.searchUsersReactComponent = null;
        return ReactDOM.unmountComponentAtNode( this.domElement );
    }

    /* 
     * @param user object {
     *
     *      hierarchyId: '"00215A9C-537A-1ED8-91D7-9D2EA1B25283"
     *      firstName: "Tom",
     *      lastName: "Lu"
     *      emailAddress: "tom.lu@tom.lu"
     * }
     *
     */
    addUser( user ) {

        this.hierarchyStorage.addUser( user );
        this.searchUsersReactComponent.showAllUsers();
    }
}

export {

    SearchUsersApp
};