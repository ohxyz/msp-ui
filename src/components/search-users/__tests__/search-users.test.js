import React from 'react';
import { SearchUsers } from '../search-users.js';
import renderer from 'react-test-renderer';
import { HierarchyStorage } from '../../core/hierarchy-storage.js';

it( 'should render with defaults', () => {

    let sap = { results: [] };
    let hs = new HierarchyStorage( sap );

    const tree = renderer
        .create( <SearchUsers storage={ hs } /> )
        .toJSON();

    expect( tree ).toMatchSnapshot();

} );

describe( 'SearchUsers Class', () => { 

    test( 'static method should throw an error', () => {
        
        expect( () => SearchUsers.getDerivedStateFromProps( {} ) ).toThrow();
    } ) 

} );