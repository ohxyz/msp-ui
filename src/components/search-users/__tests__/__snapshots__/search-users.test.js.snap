// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with defaults 1`] = `
<div
  id="search-users-section"
>
  <div
    className="search-box"
    id="seach-users-box"
  >
    <div
      className="search-box__header"
    >
      <input
        className="search-box__field"
        data-text=""
        id="search-users-field"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search users by name, department or agency"
        type="text"
      />
      <span
        className="search-box__icon"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          close
        </i>
      </span>
    </div>
  </div>
</div>
`;
